libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -3,164 +3,369 @@
+            }
+
+            public DynamicTableEntity AsTableEntity()
+            {
+                // Flatten properties
+                var properties = new Dictionary<string, EntityProperty>();
+                Flatten(properties, Entity);
+
+                // Add Type information
+                var type = Entity.GetType();
+                var typeQualifiedName = type.AssemblyQualifiedName;
+                properties.Add("__type", EntityProperty.GeneratePropertyForString(typeQualifiedName));
+
+                return new DynamicTableEntity(Key.PartitionKey, Key.RowKey)
                 {
-                    var result = await this.Table.ExecuteAsync(TableOperation.Replace(entity)).ConfigureAwait(false);
+                    ETag = ETag,
+                    Properties = properties
+                };
+            }
+
+            public static StoreItemEntity AsStoreItemEntity(TableResult tableEntity)
+            {
+                // Create instance of proper type
+                var dynamicTableEntity = (DynamicTableEntity)tableEntity.Result;
+                var type = Type.GetType(dynamicTableEntity.Properties["__type"].StringValue);
+                var properties = dynamicTableEntity.Properties;
+                var value = CreateInstaceOf(type);
+
+                // Set object properties
+                if (IsAnonymousType(type))
+                {
+                    // Anonymous Type not supported - Anonymous Types may not have setters as values are passed in the constructor
+                    // Convert to dynamic/ExpandoObject instead
+                    var expando = new ExpandoObject() as IDictionary<string, Object>;
+
+
+                    // TODO: Handle nested objects (e.g.: Parent_Child_Prop1)
+                    foreach (var prop in properties)
+                    {
+                        var propKey = prop.Key;
+                        var propValue = prop.Value.PropertyAsObject;
+                        expando[propKey] = propValue;
+                    }
+
+                    value = expando;
                 }
                 else
                 {
-                    throw new Exception("etag empty");
+                    // POCO entity
+                    TableEntity.ReadUserObject(value, properties, new OperationContext());
                 }
+
+                // IStoreItem? apply Etag
+                if (value is IStoreItem iStoreItem)
+                {
+                    iStoreItem.eTag = tableEntity.Etag;
+                }
+
+                return new StoreItemEntity(new EntityKey(dynamicTableEntity.PartitionKey, dynamicTableEntity.RowKey), value);
             }
-        }
 
-        protected class StoreItemEntity : TableEntity
-        {
-            private static JsonSerializerSettings serializationSettings = new JsonSerializerSettings()
+            private static void Flatten(IDictionary<string, EntityProperty> properties, object entity, string prefix = "")
```

Comment:
This is build into the modern SDK from Azure. Please use that, rather than custom authoring a solution. 

https://docs.microsoft.com/en-us/dotnet/api/microsoft.windowsazure.storage.table.tableentity.flatten?redirectedfrom=MSDN&view=azure-dotnet#overloads

https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Table/TableEntityAdapter.cs

Reply:
You can see how Drew did this (in a seperate PR) here:
https://github.com/Microsoft/botbuilder-dotnet/pull/376/files#diff-07ac2ab1b97373bba481752bca6ae437R123

Reply:
The TableEntity.Flatten method is being used on POCO entities, but for Anonymous Types it does not work.
There is a related discussion about supporting Anonymous Types in ChannelData (https://github.com/Microsoft/botbuilder-dotnet/issues/286).

What should be the approach for Bot/User/Conversation State ?

Reply:
I would use the stock features from the Azure SDK, and just accept the same bugs they have. Rolling our own serializatoin story is probably (overally) going to lead to more bugs and confusion.